File: L14GoldmanSachsMedium/LongestWordInDictionary.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

// Function to check if 'word' is a subsequence of 's'
bool isSubsequence(const string &s, const string &word)
{
    int i = 0, j = 0;
    while (i < s.length() && j < word.length())
    {
        if (s[i] == word[j])
        {
            j++; // Move to the next character in word
        }
        i++; // Always move in s
    }
    return j == word.length(); // If j reached word.length(), word is a subsequence
}

// Function to find the longest word in dictionary that is a subsequence of s
string findLongestWord(string s, vector<string> &dictionary)
{
    // Sort dictionary: first by length (descending), then lexicographically (ascending)
    sort(dictionary.begin(), dictionary.end(), [](const string &a, const string &b)
        { return (a.length() == b.length()) ? (a < b) : (a.length() > b.length()); });

    // Check each word in dictionary
    for (const string &word : dictionary)
    {
        if (isSubsequence(s, word))
        {                // If 'word' is a subsequence of 's'
            return word; // First valid word will be the longest due to sorting
        }
    }
    return ""; // No valid word found
}

int main()
{
    string s = "abpcplea"; // Input string
    vector<string> dictionary = {"ale", "apple", "monkey", "plea"};

    // Find and print the longest word
    cout << findLongestWord(s, dictionary) << endl;
    return 0;
}

File: L14GoldmanSachsMedium/LongestWordInDictionary.java

import java.util.*;

public class LongestWordInDictionary {

    // Function to check if 'word' is a subsequence of 's'
    private static boolean isSubsequence(String s, String word) {
        int i = 0, j = 0;
        while (i < s.length() && j < word.length()) {
            if (s.charAt(i) == word.charAt(j)) {
                j++; // Move to the next character in word
            }
            i++; // Always move in s
        }
        return j == word.length(); // If j reached word.length(), word is a subsequence
    }

    public static String findLongestWord(String s, List<String> dictionary) {
        String longestWord = "";

        // Sort dictionary: first by length (descending), then lexicographically (ascending)
        Collections.sort(dictionary, (a, b)
                -> a.length() == b.length() ? a.compareTo(b) : b.length() - a.length()
        );

        // Check each word in dictionary
        for (String word : dictionary) {
            if (isSubsequence(s, word)) { // If 'word' is a subsequence of 's'
                return word; // First valid word will be the longest due to sorting
            }
        }
        return longestWord; // No valid word found
    }

    public static void main(String[] args) {
        String s = "abpcplea"; // Input string
        List<String> dictionary = Arrays.asList("ale", "apple", "monkey", "plea");

        // Find and print the longest word
        System.out.println(findLongestWord(s, dictionary));
    }
}
